src/ompl/control/planners/sst/src/HySST.cpp

```diff
@@ -0,0 +1,567 @@
+ *     with the distribution.
+ *   * Neither the name of the University of Santa Cruz nor the names of 
+ *     its contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Authors: Beverly Xu */
+/* Adapted from: ompl/geometric/planners/src/SST.cpp by  Zakary Littlefield of Rutgers the State University of New Jersey, New Brunswick */
+
+#include "ompl/control/planners/sst/HySST.h"
+#include "ompl/base/objectives/MinimaxObjective.h"
+#include "ompl/base/objectives/MaximizeMinClearanceObjective.h"
+#include "ompl/base/objectives/PathLengthOptimizationObjective.h"
+#include "ompl/tools/config/SelfConfig.h"
+#include "ompl/base/spaces/RealVectorStateSpace.h"
+#include "ompl/base/goals/GoalState.h"
+#include "ompl/control/spaces/RealVectorControlSpace.h"
+
+namespace base = ompl::base;
+namespace tools = ompl::tools;
+namespace control = ompl::control;
+
+ompl::control::HySST::HySST(const control::SpaceInformationPtr &si_) : base::Planner(si_, "HySST")
+{
+    specs_.approximateSolutions = true;
+    siC_ = si_.get();
+    prevSolution_.clear();
+
+    addPlannerProgressProperty("best cost REAL", [this]
+                               { return std::to_string(this->prevSolutionCost_.value()); });
+}
+
+ompl::control::HySST::~HySST() {}
+
+void ompl::control::HySST::setup()
+{
+    base::Planner::setup();
+    if (!nn_)
+        nn_.reset(tools::SelfConfig::getDefaultNearestNeighbors<Motion *>(this));
+    nn_->setDistanceFunction([this](const Motion *a, const Motion *b)
+                             { return ompl::control::HySST::distanceFunc_(a->state, b->state); });
+    if (!witnesses_)
+        witnesses_.reset(tools::SelfConfig::getDefaultNearestNeighbors<Motion *>(this));
+    witnesses_->setDistanceFunction([this](const Motion *a, const Motion *b)
+                                    { return ompl::control::HySST::distanceFunc_(a->state, b->state); });
+
+    if (pdef_)
+    {
+        if (pdef_->hasOptimizationObjective())
```

Comment:
I believe `if (pdef_ && pdef_->hasOptimizationObjective())` can avoid the goto here?